src/liballoc/borrow.rs

```diff
@@ -479,3 +490,18 @@ impl<'a> AddAssign<Cow<'a, str>> for Cow<'a, str> {
         }
     }
 }
+
+#[stable(feature = "cow_str_add_assign_char", since = "1.41.0")]
+impl AddAssign<char> for Cow<'_, str> {
+    fn add_assign(&mut self, rhs: char) {
+        if let Cow::Borrowed(lhs) = *self {
+            let base_capacity = lhs.len() + rhs.len_utf8();
+            //Attempt amortized memory allocation
+            let new_optimal_size = max(base_capacity * 2, base_capacity);
```

Comment:
This doesn't look right to me -- we know base_capacity >= 0 so `max(base_capacity * 2, base_capacity)` is always the same as `base_capacity * 2`. Has it been written this way to account for integer overflow? If so, that needs to be handled differently; the overflowing multiplication would not be guaranteed to wrap around to 0.